### PR TITLE
Initial commit of Salt Package Manager

### PR DESCRIPTION
**Pull request by techhat**

### CODE REVIEW

**Comment by techhat** [comment]
Let me get some docs in before we merge this.

**Comment by cachedout** [comment]
Way awesome!

**Comment by techhat** [comment]
Docs are added, you may merge when ready.

**Comment by cachedout** [comment]
QA, we will need some test coverage for this new feature. I will file a new issue.

**Comment by nmadhok** [comment]
This is awesome! Need to start working on restructure Salt Formulas. Let's get together sometime!

**Comment by rallytime** [comment]
Double awesome!

**Comment by aboe76** [comment]
This is a nicely done :) :+1:

File: doc/topics/spm/index.rst [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+~~~~~~~~~
+The value of the ``os_family`` grain that this formula supports. This is used to
+help users know which operating system families can support this package.
+
+version
+~~~~~~~
+The version of the package. While it is up to the organization that manages this
+package, it is suggested that this version is specified in a ``YYYYMM`` format.
+For instance, if this version was released in June 2015, the package version
+should be ``201506``. If multiple released are made in a month, the ``releasee``
```
Comment (s0undt3ch) [inline doc/topics/spm/index.rst]: `releasee` -> `release`?

**Comment by aboe76** [comment]
@techhat will you add spm to setup.py so we can build it and test it?

**Comment by s0undt3ch** [comment]
@aboe76 `setup.py` now correctly installs the spm script, was that what you were referring to? Is there something else missing?

**Comment by techhat** [comment]
Sorry @aboe76, don't know why I didn't see this when you first added it.

**Comment by aboe76** [comment]
No problem you have fixed it already, was off course too quick

**Comment by iggy** [comment]
How do dependencies work (document)?

Does this support optional dependencies (like every package manager it mentions in it's influences)?

**Comment by techhat** [comment]
Hi @iggy. At the moment, hard dependencies are enforced (unless `--force`d), but not resolved. Soft/optional dependencies are not in the design, but an excellent idea.

**Comment by techhat** [comment]
@iggy, ref #25210 and #25211.

**Pull request merged**